File: engine/src/flutter/lib/web_ui/lib/src/engine/skwasm/skwasm_impl/path.dart
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

import 'dart:convert';
import 'dart:ffi';
import 'dart:typed_data';

import 'package:ui/src/engine.dart';
import 'package:ui/src/engine/skwasm/skwasm_impl.dart';
import 'package:ui/ui.dart' as ui;

enum PathDirection { clockwise, counterClockwise }

enum PathArcSize { small, large }

class SkwasmPath extends SkwasmObjectWrapper<RawPath> implements ScenePath {
  factory SkwasmPath() {
    return SkwasmPath.fromHandle(pathCreate());
  }

  factory SkwasmPath.from(SkwasmPath source) {
    return SkwasmPath.fromHandle(pathCopy(source.handle));
  }

  SkwasmPath.fromHandle(PathHandle handle) : super(handle, _registry);

  static final SkwasmFinalizationRegistry<RawPath> _registry = SkwasmFinalizationRegistry<RawPath>(
    (PathHandle handle) => pathDispose(handle),
  );

  @override
  ui.PathFillType get fillType => ui.PathFillType.values[pathGetFillType(handle)];

  @override
  set fillType(ui.PathFillType fillType) => pathSetFillType(handle, fillType.index);

  @override
  void moveTo(double x, double y) => pathMoveTo(handle, x, y);

  @override
  void relativeMoveTo(double x, double y) => pathRelativeMoveTo(handle, x, y);

  @override
  void lineTo(double x, double y) => pathLineTo(handle, x, y);

  @override
  void relativeLineTo(double x, double y) => pathRelativeLineTo(handle, x, y);

  @override
  void quadraticBezierTo(double x1, double y1, double x2, double y2) =>
      pathQuadraticBezierTo(handle, x1, y1, x2, y2);

  @override
  void relativeQuadraticBezierTo(double x1, double y1, double x2, double y2) =>
      pathRelativeQuadraticBezierTo(handle, x1, y1, x2, y2);

  @override
  void cubicTo(double x1, double y1, double x2, double y2, double x3, double y3) =>
      pathCubicTo(handle, x1, y1, x2, y2, x3, y3);

  @override
  void relativeCubicTo(double x1, double y1, double x2, double y2, double x3, double y3) =>
      pathRelativeCubicTo(handle, x1, y1, x2, y2, x3, y3);

  @override
  void conicTo(double x1, double y1, double x2, double y2, double w) =>
      pathConicTo(handle, x1, y1, x2, y2, w);

  @override
  void relativeConicTo(double x1, double y1, double x2, double y2, double w) =>
      pathRelativeConicTo(handle, x1, y1, x2, y2, w);

  @override
  void arcTo(ui.Rect rect, double startAngle, double sweepAngle, bool forceMoveTo) {
    withStackScope((StackScope s) {
      pathArcToOval(
        handle,
        s.convertRectToNative(rect),
        ui.toDegrees(startAngle),
        ui.toDegrees(sweepAngle),
        forceMoveTo,
      );
    });
  }

  @override
  void arcToPoint(
    ui.Offset arcEnd, {
    ui.Radius radius = ui.Radius.zero,
    double rotation = 0.0,
    bool largeArc = false,
    bool clockwise = true,
  }) {
    final PathArcSize arcSize = largeArc ? PathArcSize.large : PathArcSize.small;
    final PathDirection pathDirection =
        clockwise ? PathDirection.clockwise : PathDirection.counterClockwise;
    pathArcToRotated(
      handle,
      radius.x,
      radius.y,
      ui.toDegrees(rotation),
      arcSize.index,
      pathDirection.index,
      arcEnd.dx,
      arcEnd.dy,
    );
  }

  @override
  void relativeArcToPoint(
    ui.Offset arcEndDelta, {
    ui.Radius radius = ui.Radius.zero,
    double rotation = 0.0,
    bool largeArc = false,
    bool clockwise = true,
  }) {
    final PathArcSize arcSize = largeArc ? PathArcSize.large : PathArcSize.small;
    final PathDirection pathDirection =
        clockwise ? PathDirection.clockwise : PathDirection.counterClockwise;
    pathRelativeArcToRotated(
      handle,
      radius.x,
      radius.y,
      ui.toDegrees(rotation),
      arcSize.index,
      pathDirection.index,
      arcEndDelta.dx,
      arcEndDelta.dy,
    );
  }

  @override
  void addRect(ui.Rect rect) {
    withStackScope((StackScope s) {
      pathAddRect(handle, s.convertRectToNative(rect));
    });
  }

  @override
  void addOval(ui.Rect rect) {
    withStackScope((StackScope s) {
      pathAddOval(handle, s.convertRectToNative(rect));
    });
  }

  @override
  void addArc(ui.Rect rect, double startAngle, double sweepAngle) {
    withStackScope((StackScope s) {
      pathAddArc(
        handle,
        s.convertRectToNative(rect),
        ui.toDegrees(startAngle),
        ui.toDegrees(sweepAngle),
      );
    });
  }

  @override
  void addPolygon(List<ui.Offset> points, bool close) {
    withStackScope((StackScope s) {
      pathAddPolygon(handle, s.convertPointArrayToNative(points), points.length, close);
    });
  }

  @override
  void addRRect(ui.RRect rrect) {
    withStackScope((StackScope s) {
      pathAddRRect(handle, s.convertRRectToNative(rrect));
    });
  }

  @override
  void addRSuperellipse(ui.RSuperellipse rsuperellipse) {
    // TODO(dkwingsmt): Properly implement RSuperellipse on Web instead of falling
    // back to RRect.  https://github.com/flutter/flutter/issues/163718
    addRRect(rsuperellipse.toApproximateRRect());
  }

  @override
  void addPath(ui.Path path, ui.Offset offset, {Float64List? matrix4}) {
    _addPath(path, offset, false, matrix4: matrix4);
  }

  @override
  void extendWithPath(ui.Path path, ui.Offset offset, {Float64List? matrix4}) {
    _addPath(path, offset, true, matrix4: matrix4);
  }

  void _addPath(ui.Path path, ui.Offset offset, bool extend, {Float64List? matrix4}) {
    assert(path is SkwasmPath);
    withStackScope((StackScope s) {
      final Pointer<Float> convertedMatrix = s.convertMatrix4toSkMatrix(
        matrix4 ?? Matrix4.identity().toFloat64(),
      );
      convertedMatrix[2] += offset.dx;
      convertedMatrix[5] += offset.dy;
      pathAddPath(handle, (path as SkwasmPath).handle, convertedMatrix, extend);
    });
  }

  @override
  void close() => pathClose(handle);

  @override
  void reset() => pathReset(handle);

  @override
  bool contains(ui.Offset point) => pathContains(handle, point.dx, point.dy);

  @override
  ui.Path shift(ui.Offset offset) =>
      transform(Matrix4.translationValues(offset.dx, offset.dy, 0.0).toFloat64());

  @override
  ui.Path transform(Float64List matrix4) {
    return withStackScope((StackScope s) {
      final PathHandle newPathHandle = pathCopy(handle);
      pathTransform(newPathHandle, s.convertMatrix4toSkMatrix(matrix4));
      return SkwasmPath.fromHandle(newPathHandle);
    });
  }

  @override
  ui.Rect getBounds() {
    return withStackScope((StackScope s) {
      final Pointer<Float> rectBuffer = s.allocFloatArray(4);
      pathGetBounds(handle, rectBuffer);
      return s.convertRectFromNative(rectBuffer);
    });
  }

  static SkwasmPath combine(ui.PathOperation operation, SkwasmPath path1, SkwasmPath path2) =>
      SkwasmPath.fromHandle(pathCombine(operation.index, path1.handle, path2.handle));

  @override
  ui.PathMetrics computeMetrics({bool forceClosed = false}) {
    return SkwasmPathMetrics(path: this, forceClosed: forceClosed);
  }

  @override
  String toSvgString() {
    final SkStringHandle skString = pathGetSvgString(handle);
    final Pointer<Int8> buffer = skStringGetData(skString);
    final int length = skStringGetLength(skString);
    final List<int> characters = List<int>.generate(length, (int i) => buffer[i]);
    final String svgString = utf8.decode(characters);
    skStringFree(skString);
    return svgString;
  }
}

File: engine/src/flutter/lib/web_ui/skwasm/path.cpp
// Copyright 2013 The Flutter Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "export.h"
#include "helpers.h"
#include "live_objects.h"
#include "third_party/skia/include/core/SkPathBuilder.h"
#include "third_party/skia/include/core/SkString.h"
#include "third_party/skia/include/pathops/SkPathOps.h"
#include "third_party/skia/include/utils/SkParsePath.h"

using namespace Skwasm;

SKWASM_EXPORT SkPathBuilder* path_create() {
  livePathCount++;
  return new SkPathBuilder();
}

SKWASM_EXPORT void path_dispose(SkPathBuilder* path) {
  livePathCount--;
  delete path;
}

SKWASM_EXPORT SkPathBuilder* path_copy(SkPathBuilder* path) {
  livePathCount++;
  return new SkPathBuilder(path->snapshot());
}

SKWASM_EXPORT void path_setFillType(SkPathBuilder* path,
                                    SkPathFillType fillType) {
  path->setFillType(fillType);
}

SKWASM_EXPORT SkPathFillType path_getFillType(SkPathBuilder* path) {
  return path->fillType();
}

SKWASM_EXPORT void path_moveTo(SkPathBuilder* path, SkScalar x, SkScalar y) {
  path->moveTo({x, y});
}

SKWASM_EXPORT void path_relativeMoveTo(SkPathBuilder* path,
                                       SkScalar x,
                                       SkScalar y) {
  path->rMoveTo({x, y});
}

SKWASM_EXPORT void path_lineTo(SkPathBuilder* path, SkScalar x, SkScalar y) {
  path->lineTo({x, y});
}

SKWASM_EXPORT void path_relativeLineTo(SkPathBuilder* path,
                                       SkScalar x,
                                       SkScalar y) {
  path->rLineTo({x, y});
}

SKWASM_EXPORT void path_quadraticBezierTo(SkPathBuilder* path,
                                          SkScalar x1,
                                          SkScalar y1,
                                          SkScalar x2,
                                          SkScalar y2) {
  path->quadTo({x1, y1}, {x2, y2});
}

SKWASM_EXPORT void path_relativeQuadraticBezierTo(SkPathBuilder* path,
                                                  SkScalar x1,
                                                  SkScalar y1,
                                                  SkScalar x2,
                                                  SkScalar y2) {
  path->rQuadTo({x1, y1}, {x2, y2});
}

SKWASM_EXPORT void path_cubicTo(SkPathBuilder* path,
                                SkScalar x1,
                                SkScalar y1,
                                SkScalar x2,
                                SkScalar y2,
                                SkScalar x3,
                                SkScalar y3) {
  path->cubicTo({x1, y1}, {x2, y2}, {x3, y3});
}

SKWASM_EXPORT void path_relativeCubicTo(SkPathBuilder* path,
                                        SkScalar x1,
                                        SkScalar y1,
                                        SkScalar x2,
                                        SkScalar y2,
                                        SkScalar x3,
                                        SkScalar y3) {
  path->rCubicTo({x1, y1}, {x2, y2}, {x3, y3});
}

SKWASM_EXPORT void path_conicTo(SkPathBuilder* path,
                                SkScalar x1,
                                SkScalar y1,
                                SkScalar x2,
                                SkScalar y2,
                                SkScalar w) {
  path->conicTo({x1, y1}, {x2, y2}, w);
}

SKWASM_EXPORT void path_relativeConicTo(SkPathBuilder* path,
                                        SkScalar x1,
                                        SkScalar y1,
                                        SkScalar x2,
                                        SkScalar y2,
                                        SkScalar w) {
  path->rConicTo({x1, y1}, {x2, y2}, w);
}

SKWASM_EXPORT void path_arcToOval(SkPathBuilder* path,
                                  const SkRect* rect,
                                  SkScalar startAngle,
                                  SkScalar sweepAngle,
                                  bool forceMoveTo) {
  path->arcTo(*rect, startAngle, sweepAngle, forceMoveTo);
}

SKWASM_EXPORT void path_arcToRotated(SkPathBuilder* path,
                                     SkScalar rx,
                                     SkScalar ry,
                                     SkScalar xAxisRotate,
                                     SkPathBuilder::ArcSize arcSize,
                                     SkPathDirection pathDirection,
                                     SkScalar x,
                                     SkScalar y) {
  path->arcTo({rx, ry}, xAxisRotate, arcSize, pathDirection, {x, y});
}

SKWASM_EXPORT void path_relativeArcToRotated(SkPathBuilder* path,
                                             SkScalar rx,
                                             SkScalar ry,
                                             SkScalar xAxisRotate,
                                             SkPathBuilder::ArcSize arcSize,
                                             SkPathDirection pathDirection,
                                             SkScalar x,
                                             SkScalar y) {
  path->rArcTo({rx, ry}, xAxisRotate, arcSize, pathDirection, {x, y});
}

SKWASM_EXPORT void path_addRect(SkPathBuilder* path, const SkRect* rect) {
  path->addRect(*rect);
}

SKWASM_EXPORT void path_addOval(SkPathBuilder* path, const SkRect* oval) {
  path->addOval(*oval, SkPathDirection::kCW, 1);
}

SKWASM_EXPORT void path_addArc(SkPathBuilder* path,
                               const SkRect* oval,
                               SkScalar startAngle,
                               SkScalar sweepAngle) {
  path->addArc(*oval, startAngle, sweepAngle);
}

SKWASM_EXPORT void path_addPolygon(SkPathBuilder* path,
                                   const SkPoint* points,
                                   int count,
                                   bool close) {
  path->addPolygon({points, count}, close);
}

SKWASM_EXPORT void path_addRRect(SkPathBuilder* path,
                                 const SkScalar* rrectValues) {
  path->addRRect(createSkRRect(rrectValues), SkPathDirection::kCW);
}

SKWASM_EXPORT void path_addPath(SkPathBuilder* path,
                                const SkPathBuilder* other,
                                const SkScalar* matrix33,
                                SkPath::AddPathMode extendPath) {
  path->addPath(other->snapshot(), createSkMatrix(matrix33), extendPath);
}

SKWASM_EXPORT void path_close(SkPathBuilder* path) {
  path->close();
}

SKWASM_EXPORT void path_reset(SkPathBuilder* path) {
  path->reset();
}

SKWASM_EXPORT bool path_contains(SkPathBuilder* path, SkScalar x, SkScalar y) {
  return path->contains({x, y});
}

SKWASM_EXPORT void path_transform(SkPathBuilder* path,
                                  const SkScalar* matrix33) {
  path->transform(createSkMatrix(matrix33));
}

SKWASM_EXPORT void path_getBounds(SkPathBuilder* path, SkRect* rect) {
  *rect = path->computeFiniteBounds().value_or(SkRect());
}

SKWASM_EXPORT SkPathBuilder* path_combine(SkPathOp operation,
                                          const SkPathBuilder* path1,
                                          const SkPathBuilder* path2) {
  if (auto result = Op(path1->snapshot(), path2->snapshot(), operation)) {
    livePathCount++;
    SkPathBuilder* output = new SkPathBuilder(result.value());
    output->setFillType(path1->fillType());
    return output;
  } else {
    return nullptr;
  }
}

SKWASM_EXPORT SkString* path_getSvgString(SkPathBuilder* path) {
  liveStringCount++;
  SkString* string = new SkString(SkParsePath::ToSVGString(path->snapshot()));
  return string;
}
